File: test/scripts/27-zero-copy/correct.py
#!/usr/bin/env python3
"""
CORRECT: memoryview / slice - no copy, reference to same buffer.
Resume: "Use Zero-Copy Patterns"
"""
import os
import signal
import time
N = 1_000_000
CHUNK = 1000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "60"))
stop = False

def main():
    global stop
    data = bytearray(b'x' * CHUNK)
    mv = memoryview(data)
    print("CORRECT: memoryview (no copy)")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    while not stop and (time.perf_counter() - start) < DURATION_SEC:
        for _ in range(N):
            view = mv[0:100]
            data[0] = 1
        laps += 1
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/27-zero-copy/correct.js
const N = 1_000_000;
const CHUNK = 1000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '60', 10);
const data = Buffer.alloc(CHUNK, 'x');
let laps = 0;
const start = Date.now();
console.log('CORRECT: Buffer.slice (view, no copy)');
function run() {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    console.log('Stopped:', laps, 'laps');
    return;
  }
  for (let i = 0; i < N; i++) {
    const view = data.subarray(0, 100);
    data[0] = 1;
  }
  laps++;
  setImmediate(run);
}
run();
process.on('SIGINT', () => process.exit(0));
